File: backend/app/services/energy.py
from datetime import date

from app.schemas.aggregates import RestingEnergyEstimate

FORMULA_VERSION = "mifflin-st-jeor-1990"
SUPPORTED_MIN_AGE = 19
SUPPORTED_MAX_AGE = 78


def age_on(birth_date: date, on_date: date) -> int:
    return on_date.year - birth_date.year - (
        (on_date.month, on_date.day) < (birth_date.month, birth_date.day)
    )


def estimate_resting_energy(
    *,
    sex: str | None,
    birth_date: date | None,
    height_cm: float | None,
    weight_kg: float | None,
    on_date: date,
) -> RestingEnergyEstimate:
    """Return an explainable REE estimate; never silently invent profile inputs."""

    missing_fields = [
        name
        for name, value in (
            ("sex", sex),
            ("birth_date", birth_date),
            ("height_cm", height_cm),
            ("weight_kg", weight_kg),
        )
        if value is None
    ]
    if missing_fields:
        return RestingEnergyEstimate(
            status="missing_profile",
            formula=FORMULA_VERSION,
            missing_fields=missing_fields,
            note="补充性别、生日、身高和体重后可生成静息消耗估算。",
        )

    if sex not in {"female", "male"}:
        return RestingEnergyEstimate(
            status="unsupported_profile",
            formula=FORMULA_VERSION,
            missing_fields=[],
            note="当前标准公式只提供女性/男性系数，因此不对该性别选项强行生成单一数值。",
        )

    assert birth_date is not None
    assert height_cm is not None
    assert weight_kg is not None
    age_years = age_on(birth_date, on_date)
    if not SUPPORTED_MIN_AGE <= age_years <= SUPPORTED_MAX_AGE:
        return RestingEnergyEstimate(
            status="unsupported_profile",
            formula=FORMULA_VERSION,
            age_years=age_years,
            missing_fields=[],
            note=f"该公式的原始健康成人样本年龄为 {SUPPORTED_MIN_AGE}—{SUPPORTED_MAX_AGE} 岁，当前不输出估算。",
        )

    sex_constant = 5 if sex == "male" else -161
    kcal_per_day = 10 * weight_kg + 6.25 * height_cm - 5 * age_years + sex_constant
    return RestingEnergyEstimate(
        status="available",
        formula=FORMULA_VERSION,
        kcal_per_day=round(kcal_per_day, 2),
        age_years=age_years,
        missing_fields=[],
        note="静息能量消耗预测值，不是代谢测量，也不等于包含日常活动和食物热效应的 TDEE。",
    )

File: apps/mobile/src/lib/energy.ts
import type { HomeToday, JourneyProfile } from '@journey/contracts';

const FORMULA: HomeToday['resting_energy']['formula'] = 'mifflin-st-jeor-1990';

function ageOn(birthDate: string, onDate: string): number {
  const birth = birthDate.split('-').map(Number);
  const current = onDate.split('-').map(Number);
  let age = current[0]! - birth[0]!;
  if (current[1]! < birth[1]! || (current[1] === birth[1] && current[2]! < birth[2]!)) age -= 1;
  return age;
}

export function estimateRestingEnergy(
  profile: JourneyProfile,
  weightKg: number | null,
  onDate: string,
): HomeToday['resting_energy'] {
  const missing = [
    ['sex', profile.sex], ['birth_date', profile.birth_date],
    ['height_cm', profile.height_cm], ['weight_kg', weightKg],
  ].filter(([, value]) => value == null).map(([name]) => String(name));
  if (missing.length) return {
    status: 'missing_profile', kcal_per_day: null, formula: FORMULA, age_years: null,
    missing_fields: missing, note: '补充性别、生日、身高和体重后可生成静息消耗估算。',
  };
  if (profile.sex !== 'female' && profile.sex !== 'male') return {
    status: 'unsupported_profile', kcal_per_day: null, formula: FORMULA, age_years: null,
    missing_fields: [], note: '当前标准公式只提供女性/男性系数，因此不对该性别选项强行生成单一数值。',
  };
  const age = ageOn(profile.birth_date!, onDate);
  if (age < 19 || age > 78) return {
    status: 'unsupported_profile', kcal_per_day: null, formula: FORMULA, age_years: age,
    missing_fields: [], note: '该公式的原始健康成人样本年龄为 19—78 岁，当前不输出估算。',
  };
  const kcal = 10 * weightKg! + 6.25 * profile.height_cm! - 5 * age + (profile.sex === 'male' ? 5 : -161);
  return {
    status: 'available', kcal_per_day: Math.round(kcal * 100) / 100, formula: FORMULA,
    age_years: age, missing_fields: [],
    note: '静息能量消耗预测值，不是代谢测量，也不等于包含日常活动和食物热效应的 TDEE。',
  };
}
